crypto/evp/digest.c

```diff
@@ -188,18 +188,28 @@ int EVP_DigestFinalXOF(EVP_MD_CTX *ctx, unsigned char *md, size_t size)
         && ctx->digest->md_ctrl(ctx, EVP_MD_CTRL_XOF_LEN, (int)size, NULL)) {
         ret = ctx->digest->final(ctx, md);
 
-        if (ctx->digest->cleanup != NULL) {
-            ctx->digest->cleanup(ctx);
-            EVP_MD_CTX_set_flags(ctx, EVP_MD_CTX_FLAG_CLEANED);
-        }
-        OPENSSL_cleanse(ctx->md_data, ctx->digest->ctx_size);
+        /* We don't clean up here, since it's okay to squeeze an XOF more
+         * than once. */
     } else {
-        EVPerr(EVP_F_EVP_DIGESTFINALXOF, EVP_R_NOT_XOF_OR_INVALID_LENGTH);
+        EVPerr(EVP_F_EVP_DIGESTSQUEEZEXOF, EVP_R_NOT_XOF_OR_INVALID_LENGTH);
     }
 
     return ret;
 }
 
+int EVP_DigestFinalXOF(EVP_MD_CTX *ctx, unsigned char *md, size_t size)
+{
+    int ret = EVP_DigestSqueezeXOF(ctx, md, size);
+
+    if (ctx->digest->cleanup != NULL) {
+        ctx->digest->cleanup(ctx);
+        EVP_MD_CTX_set_flags(ctx, EVP_MD_CTX_FLAG_CLEANED);
```

Comment:
Shouldn't this flag be set regardless, even if `ctx->digest->cleanup == NULL`?

Reply:
Possibly, but I am pretty sure that I didn't change that.  It's been that way ever since OpenSSL 0.9.7, and I'm not sure I should change it in this ticket.

Reply:
Ah ok